### PR TITLE
upgrade postcss

### PR DESCRIPTION
**Pull request by TrySound**
Ref https://github.com/postcss/postcss/releases/tag/8.0.0

Looks like after all we have to drop node 8 support.

### CODE REVIEW

**Comment by TrySound** [comment]
cc @bholloway

**Comment by bholloway** [comment]
I read "changes to API" and thought the worst, thanks for showing this change is not too painful. 👍 

Although I am not sure how to time this. V4 is overdue and it would be nice to ship something compatible with node 8 simply for all those poor souls stuck with old projects.

That said I think keeping on top of postcss updates is a good thing so it would be a quick followup if delayed.

**Comment by TrySound** [comment]
Yeah, probably v4 can be released with just adjust loader upgrade and then v5 with new schema utils and postcss.

**Comment by bholloway** [comment]
Webpack 5 requires 10 as minimum and just looking at the v5 project it already has some related tasks

* bump node to `>=10`.
* drop support for Webpack 2 and Webpack 3

So I will add this to v5.

**Comment by ZeProf2Code** [comment]
Is there any workaround atm?

**Comment by TrySound** [comment]
@ZeProf2Code Why? Do you have problem with postcss 7?

**Comment by ZeProf2Code** [comment]
I want to upgrade to TailwindCSS 2 and I need PostCSS 8.

**Comment by TrySound** [comment]
AFAIK in this package postcss is used only internally. It should not affect upgrade on your side.

**Comment by brian-kephart** [comment]
I'm having the same difficulty as @ZeProf2Code when attempting to upgrade PostCSS to version 8. Webpack gets a ton of compile errors pointing back to this package. It seems like PostCSS 7/8 don't coexist peacefully in this context.

**Comment by TrySound** [comment]
Which errors? Might be unrelated problem.

**Comment by dvdknaap** [comment]
Is there any update because more and more packages are requiring postcss 8 ?

**Comment by dvdknaap** [comment]
@bholloway when i force the version 4 and postcss 8 everything is working on my end

`yarn add resolve-url-loader@4.0.0-alpha.3 --dev`
`yarn add postcss@^8.0.0 --dev`

**Comment by bholloway** [comment]
@dvdknaap I'm thinking I could probably put a semver comparitor in here that allows `postcss@8`. Something like
```
"postcss": ">=7.0.35 <9",
```

I'm not sure that it would drop support for node 8 per sec, since anyone on node 8 could still use `postcss@7`.

Does that sound like it would work?

**Comment by jmfrancois** [comment]
It could also be a peer dependencies ? so the project decide the version it uses and you try to ensure compatiblity with as many versions as you want on your side ?

On our side we need to upgrade to postcss 8 because of security fix provided in last release (8.2.10)

> Fixed ReDoS vulnerabilities in source map parsing.

**Comment by ErideonTech** [comment]
https://nvd.nist.gov/vuln/detail/CVE-2021-23368

resolve-url-loader is a package used by Angular (compiler only), hence flags as a security issue for us
Thanks for updating

**Comment by bholloway** [comment]
Okay let me see what we can do.

~Please lets move to the pinned discussion issue to discuss further.~
~EDIT - my mistake there is no discussion issue.~
EDIT - #198 is now the discussion issue.

In the mean time please consider the workaround @dvdknaap suggested above.

**Comment by bholloway** [comment]
For some reason this is failing hard in E2E tests. As soon as I work out the issue I will merge the PR. If there is any additional work I will do that on the `v5` branch rather than asking for more changes here.

@TrySound et. al. FYI for `v5` I intend to require `node@>=12` and to push all the dependencies to the latest.

File: packages/resolve-url-loader/lib/engine/postcss.js [hunk 1]
```diff
@@ -28,37 +28,12 @@ function process(sourceFile, sourceContent, params) {
...
-    };
-
-    /**
-     * Process a declaration from the syntax tree.
-     * @param declaration
-     */
-    function eachDeclaration(declaration) {
+  const postcssPlugin = {
+    postcssPlugin: 'postcss-resolve-url',
+    Declaration: (declaration) => {
```
Comment (bholloway) [inline packages/resolve-url-loader/lib/engine/postcss.js:36]: This code has been locking up E2E tests. The test eventually crashes Node with "out of memory" error.

Finally worked out from [this migration guide](https://evilmartians.com/chronicles/postcss-8-plugin-migration) that...
> plugins will re-visit all changed or added nodes. You should check if your transformations were already applied and if that is the case—ignore those nodes.

So here the code needs to somehow keep track of visited declarations.
```
() => {
  postcssPlugin: 'postcss-resolve-url',
  prepare: () => {
    const visited = new Set();
    return {
      Declaration: (declaration) => {
        ...
      }
    };
  }
};
```

I will fix this and make some other tweaks post-merge.

**Review by bholloway — APPROVED** [review]

**Pull request merged**